### DISCUSSION TITLE
Get rid of external crypto libs?

### DISCUSSION THREAD

**Discussion by depler** [transmission/transmission#3567] [Ideas]
As I see it in code, transmission uses `SHA1`, `SHA256` and `SSL/TLS` crypto routines. The last one is needed only for communication with HTTPS trackers, which is very rare case - most of them are `UDP` or `HTTP`.  

So it seems possible to move `SHA` implementation as raw source code (like it is already done for `arc4` algorithm). That will allow to remove `OpenSSL/WolfSSL` dependencies in optional manner and build extremely tiny executable for certain platforms. @ckerr your thoughts?

**Comment by ckerr** [comment]
@depler I'm not sure how much this will actually simplify everything? AFAIK we still need https://github.com/transmission/transmission/commit/f3968b7708517bc61f72cd35b12a45d7a543cb07 for https://github.com/transmission/transmission/issues/446

**Reply by depler** [reply]
This is not about simplifying overall. The idea is to have an ability to build transmission binary without SSL routines (that means without support of HTTPS trackers) - to reduce binary size and remove functionality which is not needed in most cases.

**Reply by Petrprogs** [reply]
I uphold! Maybe add this option to the cmake file

**Comment by ckerr** [comment]
But yes agreed about the complexity involved in having four different crypto libs :dizzy_face:

**Reply by depler** [reply]
As for different crypto libs - there is an ability to use `WolfSSL` only as third-party component for crypto routines. It is cross-platform and quite easy to compile, contains only `c` code. Actually I've done integrating into my project and going to replace OpenSSL with it: https://github.com/depler/transmission-vs/tree/master/wolfssl_src. Binary is only ~3Mb, comparing to ~6Mb when linking with `OpenSSL`.

Probably this approach would be nice to have in main repo. It will need to add CURL as third-party code component also, but then we can safely remove support of any other crypto libs.
